### ISSUE TITLE
azure-sdk-for-ruby tests hanging in Ruby v2.0.0, v2.1.5, v2.2.0, v2.3.0

### ISSUE THREAD

**Issue by floehopper** [freerange/mocha#276]
This was originally reported by @sarangan12 in https://github.com/freerange/mocha/issues/272#issuecomment-256505558. However, I'm not convinced it's the same problem as was originally reported in #272, because in this case I can use ctrl-C to abort the tests. In #272, I had to use ctrl-Z and manually kill the process.

It doesn't look as if the code which caused the [build failures](https://travis-ci.org/Azure/azure-sdk-for-ruby/builds/170262279) exists any more, but I think I've been able to reproduce the problem as follows:
- `rbenv local 2.3.0`
- `git clone git@github.com:Azure/azure-sdk-for-ruby.git`
- `git checkout b1314c7b0c6ece809d979dc6d83abf0d62af3419`
- Change the mocha dependency in `azure.gemspec` to `~> 1.0`
- `bundle install`

```
$ bundle exec ruby -Itest test/unit/virtual_machine_management/virtual_machine_management_service_test.rb -v --seed 4895
Run options: -v --seed 4895

# Running:

Azure::VirtualMachineManagement::VirtualMachineManagementService::#get_virtual_machine#test_0001_return nil if virtual machine and cloud server does not exist  = 0.01 s = .
Azure::VirtualMachineManagement::VirtualMachineManagementService::#get_virtual_machine#test_0003_return virtual machine instance if virtual machine name and cloud server name are valid  = 0.00 s = .
Azure::VirtualMachineManagement::VirtualMachineManagementService::#get_virtual_machine#test_0002_return nil if virtual machine or cloud server does not exist  = 0.00 s = .
Azure::VirtualMachineManagement::VirtualMachineManagementService::#get_image#test_0002_errors if the virtual machine image does not exist = 0.01 s = .

# The tests hang here, but I can use ctrl-C to abort.
```

**Comment by floehopper** [comment]
A simpler way to get an appropriate version of the azure-sdk-for-ruby code to reproduce the problem is to checkout the [asm branch](https://github.com/Azure/azure-sdk-for-ruby/tree/asm) rather than the specific commit SHA that I suggested in my previous comment.

**Comment by floehopper** [comment]
I used [GDB](https://www.gnu.org/software/gdb/) to generate a Ruby backtrace at the point where the tests were hanging. The test is hanging [in the subject block on line 21](https://github.com/Azure/azure-sdk-for-ruby/blob/b1314c7b0c6ece809d979dc6d83abf0d62af3419/test/unit/virtual_machine_management/virtual_machine_management_service_test.rb#L21).

The problem seems to be related to [line 413 of the test](https://github.com/Azure/azure-sdk-for-ruby/blob/b1314c7b0c6ece809d979dc6d83abf0d62af3419/test/unit/virtual_machine_management/virtual_machine_management_service_test.rb#L413) where all the private instance methods of the subject class, `Azure::VirtualMachineManagementService` are made public.

I can stop the tests from hanging by any one of the following:

* Only make the `get_image` method public - this seems to be sufficient to pass the test
* Only make the private instance methods of the class itself public, i.e. not its ancestors (by calling `private_instance_methods(false)`)
* Defining an `initialize` method on the subject class which just calls `super`

Although the test should not be hanging, line 413 of the test seems like a really dangerous thing to do, because it makes all of the private methods public for *all* tests after that point. This is because the class is effectively a global variable and so you are not protected by the tests creating a new instance each time.

@sarangan12: I will continue to investigate why the test is hanging, but in the meantime, I'd suggest you work around the problem by using one of my suggestions above.

**Comment by floehopper** [comment]
I've now narrowed the problem down to the simplest scenario:

```ruby
module Foo
end

class Bar
  prepend Foo

  public :initialize
end

p Bar.new # => hangs
```

The newest versions of Mocha prepends a module onto the `Azure::VirtualMachineManagementService` class when the `any_instance` stub is setup [in lines 181-183 of the test](https://github.com/Azure/azure-sdk-for-ruby/blob/b1314c7b0c6ece809d979dc6d83abf0d62af3419/test/unit/virtual_machine_management/virtual_machine_management_service_test.rb#L181-L183).

[Line 413 of the test](https://github.com/Azure/azure-sdk-for-ruby/blob/b1314c7b0c6ece809d979dc6d83abf0d62af3419/test/unit/virtual_machine_management/virtual_machine_management_service_test.rb#L413) then sets the visibility of the `initialize` method to public as described in my previous comment.

Then when `Azure::VirtualMachineManagementService.new` is called [in the subject block on line 21](https://github.com/Azure/azure-sdk-for-ruby/blob/b1314c7b0c6ece809d979dc6d83abf0d62af3419/test/unit/virtual_machine_management/virtual_machine_management_service_test.rb#L21), Ruby hangs.

I have created [a Ruby bug report for the issue](https://bugs.ruby-lang.org/issues/12876).

I'm not sure that there is much we can do in Mocha other than reverting back to not using a prepended module for stubbing. @chrisroos: Do you have any thoughts?

**Comment by floehopper** [comment]
The bug in Ruby has been fixed and released in [Ruby v2.4.0](http://svn.ruby-lang.org/repos/ruby/tags/v2_4_0/doc/ChangeLog-2.4.0) and back-ported & released in [Ruby v2.3.3](http://svn.ruby-lang.org/repos/ruby/tags/v2_3_3/ChangeLog).

**Comment by floehopper** [comment]
I've added [1] a known issue to the Mocha README. I don't think there's anything else to do here, so I'm going to close this issue.

[1]: 3e202adb0a647e7e830d620864dd0f90488ab58a
